### PR TITLE
Add HyperRAM buffering to analyzer

### PR DESCRIPTION
**Pull request by martinling**
This PR is stacked on #104 and tries to add HyperRAM buffering to the analyzer.

The code is taken from [this branch](https://github.com/mndza/cynthion-analyzer/tree/analyzer-refactor) by @mndza. The integration has been adapted to apply on top of #104, with some additional testing and verification. The `Stream16to8`, `StreamFIFO` and `HyperRAMPacketFIFO` modules are taken unmodified from that branch, and were known to work there.

I have broken down the changes into stages to help isolate the remaining issues. Tests pass at all but the final two steps.

1. Remove a couple of confusing and unused signals. :heavy_check_mark: 
2. Simplify and improve the `USBAnalyzer` unit tests. :heavy_check_mark: 
3. Use a 16-bit read port for the FIFO, followed by a `Stream16to8` conversion. :heavy_check_mark:
4. Move the `Stream16to8` conversion into the top level module. :heavy_check_mark:
5. Move the FIFO read port into the `sync` clock domain, and use a `StreamFIFO` to cross back to `usb`. :heavy_multiplication_x: 
6. Insert `HyperRAMPacketFIFO` into the output pipeline. :heavy_multiplication_x:

The problems appear at step 5, and are caught by the unit tests, e.g:

```
  File "/home/martin/repos/cynthion-analyzer/dependencies/cynthion/cynthion/python/src/gateware/analyzer/analyzer.py", line 362, in expect_data
    self.assertEqual(received_data, expected_data)
AssertionError: Lists differ: [0, 10, 0, 10, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7] != [0, 10, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

First differing element 3:
10
0

- [0, 10, 0, 10, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7]
?            -      ------

+ [0, 10, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
?                                     ++++++
```

Basically the output is almost right, but slightly out of sync.

### CODE REVIEW

**Comment by martinling** [comment]
Updated, I got a bit further.

The tests get the correct data now, but the tests still fail because the `valid` signal is not deasserted at the end.

**Comment by martinling** [comment]
Updated. Thanks to @mndza spotting a remaining misuse of the `data_commit` signal, the unit tests all pass now. We're just trying to get it to pass timing now.

**Comment by martinling** [comment]
Updated again. Current status:

- By trying enough different `--seed` values in `nextpnr`, it's possible to get this to meet timing and build successfully.
- The penultimate step (moving FIFO read into `sync` domain) fails the HITL test of FS capture, but passes the HS test.
- I added a unit test which simulates a Full Speed capture (i.e. bytes arrive non-contiguously from the PHY); it passes.
- The final step (adding the HyperRAM buffer) doesn't work at all; received data seems to be random garbage.

**Comment by martinling** [comment]
Updated once more, @mndza fixed the `HyperRAMPacketFIFO`!

Remaining issues:
- FS capture is still broken at the penultimate commit: it works in simulation, but enumeration is not properly captured in the HITL test.
- Timing is still tight enough in the `sync` domain that it can take trying a few different random seeds to build successfully.

**Comment by martinling** [comment]
Thanks to more fantastic work from @mndza this is now good to go!

File: cynthion/python/src/gateware/analyzer/analyzer.py [hunk 1]
```diff
@@ -546,8 +561,13 @@ def instantiate_dut(self):
         m = Module()
         m.submodules.translator = self.translator = UTMITranslator(ulpi=self.ulpi, handle_clocking=False)
         m.submodules.analyzer   = self.analyzer   = USBAnalyzer(utmi_interface=self.translator, mem_depth=128)
-        m.submodules.s16to8 = s16to8 = DomainRenamer("usb")(Stream16to8())
-        m.d.comb += s16to8.input.stream_eq(self.analyzer.stream)
+        reset_on_start = ResetInserter(self.analyzer.discarding)
+        m.submodules.clk_conv = clk_conv = StreamSyncUsbConverter()
```
Comment (mndza) [inline cynthion/python/src/gateware/analyzer/analyzer.py:565]: I would add `reset_on_start` here too: the output of this block is registered, so it can buffer single words while `analyzer.discarding` is active.

File: cynthion/python/src/gateware/analyzer/top.py [hunk 2]
```diff
@@ -328,8 +328,11 @@ def elaborate(self, platform):
         # Create a USB analyzer.
         m.submodules.analyzer = analyzer = USBAnalyzer(utmi_interface=utmi)
 
-        # Add a special stream clock converter for 'sync' to 'usb' crossing.
+        # Follow this with a HyperRAM FIFO for additional buffering.
         reset_on_start = ResetInserter(analyzer.discarding)
+        m.submodules.psram_fifo = psram_fifo = reset_on_start(HyperRAMPacketFIFO())
```
Comment (mndza) [inline cynthion/python/src/gateware/analyzer/top.py:333]: We should set `out_fifo_depth` to some big enough value to contain bursts. If we don't, all read bursts will be 2 words at most.

Comment (martinling) [reply cynthion/python/src/gateware/analyzer/top.py]: I set it to 4096, which is more than needed - but I think it makes sense to have a bit of capacity there, to make sure the HyperRAM still gets to do large bursts on the read side even if the host's endpoint reading is a bit jittery.

**Review by mndza — APPROVED** [review]
LGTM!
